Reproject NDCube Documentation (#439)
sunpy/ndcube: docs/reproject.rst

.. _reproject:

=======================
Reprojecting ND Objects
=======================

Reprojecting allows you to transform your ND Objects to use coordinates described by another WCS object.

> Users can use this feature to resample the resolution of ND Objects by providing an appropriate target WCS. It also enables them to remove tiny differences between similar `~ndcube.NDCube` objects to get them onto the same grid.

Reviewer: ```suggestion
Using this feature it is possible to regrid ND Objects by providing an appropriate target WCS, for operations such as resampling or alignment. It also enables removing differences between similar `~ndcube.NDCube` objects to get them onto the same grid.
```

.. _cube_reproject:

Reprojecting an NDCube
======================

Reprojecting returns a new `~ndcube.NDCube` object that has been transformed to use the provided ``target_wcs``.

> The ``target_wcs`` must be compatible with the WCS that is already associated with your `~ndcube.NDCube`. This means that it should represent the same physical axes and in the same order.

Reviewer: Mostly you have done this but rst files should have one sentence per line:

```suggestion
The ``target_wcs`` must be compatible with the WCS that is already associated with your `~ndcube.NDCube`. 
This means that it should represent the same physical axes and in the same order.
```

Author: Ah okay

.. _reproject:

=======================
Reprojecting ND Objects
=======================

> Reprojecting allows you to transform your ND Objects to use coordinates described by another WCS object.

Reviewer 2: ```suggestion
Reprojecting allows you to transform your ND Objects to a coordinate grid described by another WCS object.
```

Using this feature it is possible to regrid ND Objects by providing an appropriate target WCS, for operations such as resampling or alignment.

> It also enables removing differences between similar `~ndcube.NDCube` objects to get them onto the same grid.

Reviewer 2: ```suggestion
It also enables putting similar `~ndcube.NDCube` objects onto the same grid for more direct comparison.
```

  >>> # Define data array.
  >>> data = np.random.rand(4, 4, 5)

  >>> # Define WCS transformations in an astropy WCS object.
  >>> wcs = astropy.wcs.WCS(naxis=3)
  >>> wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
  >>> wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
  >>> wcs.wcs.cdelt = 0.2, 0.5, 0.4
  >>> wcs.wcs.crpix = 0, 2, 2
  >>> wcs.wcs.crval = 10, 0.5, 1

  >>> # Instantiate NDCube with supporting data.
  >>> my_cube = NDCube(data, wcs=wcs)

.. code-block:: python

  >>> # Create a target WCS object with new transformations.
  >>> target_wcs = astropy.wcs.WCS(naxis=3)
  >>> target_wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
  >>> target_wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
  >>> target_wcs.wcs.cdelt = 0.1, 0.5, 0.4
  >>> target_wcs.wcs.crpix = 0, 2, 2
  >>> target_wcs.wcs.crval = 10, 0.5, 1

  >>> reprojected_cube = my_cube.reproject_to(target_wcs=target_wcs, shape_out=(4, 4, 10))

In the above example, the ``CDELT1`` parameter of the ``target_wcs`` was modified to 0.1 (from 0.2).
This allowed us to upsample the wavelength axis by a factor of 2.
Accordingly, the shape of the data was updated to ``(4, 4, 10)`` from the initial shape ``(4, 4, 5)``.

> The wavelength axis goes last since the pixel shape and WCS shape are represented in opposite directions.

Reviewer 2: ```suggestion
The wavelength axis goes last since the aray shape and WCS shape are represented in opposite directions.
```
